### PR TITLE
Curve-based animation

### PR DESCRIPTION
**Pull request by mweatherley**
# Objective

This PR extends and reworks the material from #15282 by allowing arbitrary curves to be used by the animation system to animate arbitrary properties. The goals of this work are to:
- Allow far greater flexibility in how animations are allowed to be defined in order to be used with `bevy_animation`.
- Delegate responsibility over keyframe interpolation to `bevy_math` and the `Curve` libraries and reduce reliance on keyframes in animation definitions generally.
- Move away from allowing the glTF spec to completely define animations on a mechanical level.

### Overview

At a high level, curves have been incorporated into the animation system using the `AnimationCurve` trait (closely related to what was `Keyframes`). From the top down:

1. In `animate_targets`, animations are driven by `VariableCurve`, which is now a thin wrapper around a `Box<dyn AnimationCurve>`.
2. `AnimationCurve` is something built out of a `Curve`, and it tells the animation system how to use the curve's output to actually mutate component properties. The trait looks like this:
```rust
/// A low-level trait that provides control over how curves are actually applied to entities
/// by the animation system.
///
/// Typically, this will not need to be implemented manually, since it is automatically
/// implemented by [`AnimatableCurve`] and other curves used by the animation system
/// (e.g. those that animate parts of transforms or morph weights). However, this can be
/// implemented manually when `AnimatableCurve` is not sufficiently expressive.
///
/// In many respects, this behaves like a type-erased form of [`Curve`], where the output
/// type of the curve is remembered only in the components that are mutated in the
/// implementation of [`apply`].
///
/// [`apply`]: AnimationCurve::apply
pub trait AnimationCurve: Reflect + Debug + Send + Sync {
    /// Returns a boxed clone of this value.
    fn clone_value(&self) -> Box<dyn AnimationCurve>;

    /// The range of times for which this animation is defined.
    fn domain(&self) -> Interval;

    /// Write the value of sampling this curve at time `t` into `transform` or `entity`,
    /// as appropriate, interpolating between the existing value and the sampled value
    /// using the given `weight`.
    fn apply<'a>(
        &self,
        t: f32,
        transform: Option<Mut<'a, Transform>>,
        entity: EntityMutExcept<'a, (Transform, AnimationPlayer, Handle<AnimationGraph>)>,
        weight: f32,
    ) -> Result<(), AnimationEvaluationError>;
}
```
3. The conversion process from a `Curve` to an `AnimationCurve` involves using wrappers which communicate the intent to animate a particular property. For example, here is `TranslationCurve`, which wraps a `Curve<Vec3>` and uses it to animate `Transform::translation`:
```rust
/// This type allows a curve valued in `Vec3` to become an [`AnimationCurve`] that animates
/// the translation component of a transform.
pub struct TranslationCurve<C>(pub C);
```

### Animatable Properties

The `AnimatableProperty` trait survives in the transition, and it can be used to allow curves to animate arbitrary component properties. The updated documentation for `AnimatableProperty` explains this process:
<details>
  <summary>Expand AnimatableProperty example</summary

An `AnimatableProperty` is a value on a component that Bevy can animate.

You can implement this trait on a unit struct in order to support animating
custom components other than transforms and morph weights. Use that type in
conjunction with `AnimatableCurve` (and perhaps `AnimatableKeyframeCurve`
to define the animation itself). For example, in order to animate font size of a
text section from 24 pt. to 80 pt., you might use:

```rust
#[derive(Reflect)]
struct FontSizeProperty;

impl AnimatableProperty for FontSizeProperty {
    type Component = Text;
    type Property = f32;
    fn get_mut(component: &mut Self::Component) -> Option<&mut Self::Property> {
        Some(&mut component.sections.get_mut(0)?.style.font_size)
    }
}
```

You can then create an `AnimationClip` to animate this property like so:

```rust
let mut animation_clip = AnimationClip::default();
animation_clip.add_curve_to_target(
    animation_target_id,
    AnimatableKeyframeCurve::new(
        [
            (0.0, 24.0),
            (1.0, 80.0),
        ]
    )
    .map(AnimatableCurve::<FontSizeProperty, _>::from_curve)
    .expect("Failed to create font size curve")
);
```

Here, the use of `AnimatableKeyframeCurve` creates a curve out of the given keyframe time-value
pairs, using the `Animatable` implementation of `f32` to interpolate between them. The
invocation of `AnimatableCurve::from_curve` with `FontSizeProperty` indicates that the `f32`
output from that curve is to be used to animate the font size of a `Text` component (as
configured above).

</details>

### glTF Loading

glTF animations are now loaded into `Curve` types of various kinds, depending on what is being animated and what interpolation mode is being used. Those types get wrapped into and converted into `Box<dyn AnimationCurve>` and shoved inside of a `VariableCurve` just like everybody else.

### Morph Weights

There is an `IterableCurve` abstraction which allows sampling these from a contiguous buffer without allocating. Its only reason for existing is that Rust disallows you from naming function types, otherwise we would just use `Curve` with an iterator output type. (The iterator involves `Map`, and the name of the function type would have to be able to be named, but it is not.)

A `WeightsCurve` adaptor turns an `IterableCurve` into an `AnimationCurve`, so it behaves like everything else in that regard.

## Testing

Tested by running existing animation examples. Interpolation logic has had additional tests added within the `Curve` API to replace the tests in `bevy_animation`. Some kinds of out-of-bounds errors have become impossible.

Performance testing on `many_foxes` (`animate_targets`) suggests that performance is very similar to the existing implementation. Here are a couple trace histograms across different runs (yellow is this branch, red is main).
Screenshot 2024-09-27 at 9 41 50 AM
Screenshot 2024-09-27 at 9 45 18 AM

---

## Migration Guide

Most user code that does not directly deal with `AnimationClip` and `VariableCurve` will not need to be changed. On the other hand, `VariableCurve` has been completely overhauled. If you were previously defining animation curves in code using keyframes, you will need to migrate that code to use curve constructors instead. For example, a rotation animation defined using keyframes and added to an animation clip like this:
```rust
animation_clip.add_curve_to_target(
    animation_target_id,
    VariableCurve {
        keyframe_timestamps: vec![0.0, 1.0, 2.0, 3.0, 4.0],
        keyframes: Keyframes::Rotation(vec![
            Quat::IDENTITY,
            Quat::from_axis_angle(Vec3::Y, PI / 2.),
            Quat::from_axis_angle(Vec3::Y, PI / 2. * 2.),
            Quat::from_axis_angle(Vec3::Y, PI / 2. * 3.),
            Quat::IDENTITY,
        ]),
        interpolation: Interpolation::Linear,
    },
);
```

would now be added like this:
```rust
animation_clip.add_curve_to_target(
    animation_target_id,
    AnimatableKeyframeCurve::new([0.0, 1.0, 2.0, 3.0, 4.0].into_iter().zip([
        Quat::IDENTITY,
        Quat::from_axis_angle(Vec3::Y, PI / 2.),
        Quat::from_axis_angle(Vec3::Y, PI / 2. * 2.),
        Quat::from_axis_angle(Vec3::Y, PI / 2. * 3.),
        Quat::IDENTITY,
    ]))
    .map(RotationCurve)
    .expect("Failed to build rotation curve"),
);
```

Note that the interface of `AnimationClip::add_curve_to_target` has also changed (as this example shows, if subtly), and now takes its curve input as an `impl AnimationCurve`. If you need to add a `VariableCurve` directly, a new method `add_variable_curve_to_target` accommodates that (and serves as a one-to-one migration in this regard).

### For reviewers

The diff is pretty big, and the structure of some of the changes might not be super-obvious:
- `keyframes.rs` became `animation_curves.rs`, and `AnimationCurve` is based heavily on `Keyframes`, with the adaptors also largely following suite.
- The Curve API adaptor structs were moved from `bevy_math::curve::mod` into their own module `adaptors`. There are no functional changes to how these adaptors work; this is just to make room for the specialized reflection implementations since `mod.rs` was getting kind of cramped.
- The new module `gltf_curves` holds the additional curve constructions that are needed by the glTF loader. Note that the loader uses a mix of these and off-the-shelf `bevy_math` curve stuff.
- `animatable.rs` no longer holds logic related to keyframe interpolation, which is now delegated to the existing abstractions in `bevy_math::curve::cores`.

### CODE REVIEW

**Review by NthTensor — APPROVED** [review]
Skimmed this over and it looks fantastic! We probably do need to profile this a bit before merge just because animation perf tends to be so sensitive to small changes, so consider this a conditional approval. I'm also getting the tracy mismatch error, I'll post if I figure out out.

**Comment by mweatherley** [comment]
Added some tracing results.

File: crates/bevy_animation/src/animation_curves.rs
Comment (mweatherley) [reply crates/bevy_animation/src/animation_curves.rs:183]: Not really; mostly this is just a helpful grouping of constraints to avoid repeating all these trait bounds, but even in principle, `T` could still basically be anything, as long as the `AnimationCurve` implementation knows how to turn its output into a component mutation.

File: crates/bevy_animation/src/animation_curves.rs
Comment (IQuick143) [inline crates/bevy_animation/src/animation_curves.rs:289]: I wonder, why is the function signature like this?
Can't Transform be lumped into the branch for any other component?
Is this for perf reasons?

Comment (mweatherley) [reply crates/bevy_animation/src/animation_curves.rs:289]: As far as I know it's for performance reasons that I don't fully grasp. :)

Comment (pcwalton) [reply crates/bevy_animation/src/animation_curves.rs:289]: Yes, it's for performance reasons. `EntityMut` has to do access checks on its components every time you access them, while passing `Transform` like this has no access checks.

File: crates/bevy_animation/src/gltf_curves.rs
Comment (pcwalton) [inline crates/bevy_animation/src/gltf_curves.rs:272]: Not sure if this should be `#[inline]` as it's rather big.

File: crates/bevy_math/src/curve/adaptors.rs
Comment (pcwalton) [inline crates/bevy_math/src/curve/adaptors.rs:95]: Do you need to print the phantom data here?

File: crates/bevy_animation/src/gltf_curves.rs
Comment (pcwalton) [inline crates/bevy_animation/src/gltf_curves.rs:400]: I suspect this would be clearer, more numerically stable, and possibly faster with de Casteljau subdivision (repeated nested lerps), the way I had it before. At the very least this should be using [Horner's method](http://en.wikipedia.org/wiki/Horner's_method) to evaluate the polynomials faster.

File: crates/bevy_animation/src/animation_curves.rs
Comment (pcwalton) [inline crates/bevy_animation/src/animation_curves.rs:461]: We should have a TODO or an issue filed to SIMDify this at some point.

**Review by pcwalton — CHANGES_REQUESTED** [review]
Looks generally fine and definitely a nice improvement in terms of generality. I had a few questions but I'll approve once that's done.

File: crates/bevy_animation/src/gltf_curves.rs
Comment (IQuick143) [reply crates/bevy_animation/src/gltf_curves.rs:400]: If this method was actually called often (with the same coefficients), what you should do is pre-compute the vectors A,B,C,D, such that this method equals `A + tB + t^2 C + t^3 D`, and then eval this polynomial with Horner's method. (Like we do in the CubicSegment.)

File: crates/bevy_animation/src/animation_curves.rs
Comment (mweatherley) [reply crates/bevy_animation/src/animation_curves.rs:191]: Yeah, sorry about that. The reason for this is twofold:
1. The `Reflect` derive macro puts `FromReflect` bounds on everything. This effectively prevents some `'static` curves from being used by the animation system just because they can never be `FromReflect` (e.g. because they contain function data).
2. We still want to derive `FromReflect` in cases where the inner curve is `FromReflect`.

So the way this is currently addressed is that we opt out of deriving `FromReflect` in the `#[derive(Reflect)]` macro (so that even if the inner curve is not `FromReflect`, the whole thing can be `Reflect`) and then derive it separately with `#[derive(FromReflect)]`.

File: crates/bevy_math/src/curve/adaptors.rs
Comment (mweatherley) [reply crates/bevy_math/src/curve/adaptors.rs:95]: Probably not! Removed.

File: crates/bevy_animation/src/gltf_curves.rs
Comment (mweatherley) [reply crates/bevy_animation/src/gltf_curves.rs:400]: Yeah, this being the old version is basically a merge artifact, since I moved the old code verbatim out of `lib` into what is now `gltf_curves`, while you moved it into `animatable` and changed it. I don't have any attachment to the old version at all, so we could just axe it if we're confident the new one is better.

Comment (pcwalton) [reply crates/bevy_animation/src/gltf_curves.rs:400]: Here's a proposal:
```rust
let coeffs = (vec4(2.0, 1.0, -2.0, 1.0) * lerp + vec4(-3.0, -2.0, 3.0, -1.0) * lerp;
value_start * (coeffs.x + 1.0)
+ tangent_out_start * step_duration * lerp * coeffs.y
+ value_end * coeffs.z
+ tangent_in_end * coeffs.w
```

I doubt you can go much faster than this.

File: crates/bevy_animation/src/animation_curves.rs
Comment (mweatherley) [reply crates/bevy_animation/src/animation_curves.rs:181]: I renamed it for now to `AnimationCompatibleCurve`, which is verbose but at least conveys the intention better, I think.

File: crates/bevy_animation/src/animation_curves.rs
Comment (mweatherley) [reply crates/bevy_animation/src/animation_curves.rs:455]: This is a good idea. I'll submit an issue for it when this is merged.

File: crates/bevy_animation/src/gltf_curves.rs
Comment (mweatherley) [reply crates/bevy_animation/src/gltf_curves.rs]: I think it should be like this, if I'm not mistaken (I think it was missing a couple lerps and a +1):
```rust
let coeffs = (vec4(2.0, 1.0, -2.0, 1.0) * lerp + vec4(-3.0, -2.0, 3.0, -1.0)) * lerp;
value_start * (coeffs.x * lerp + 1.0)
    + tangent_out_start * step_duration * lerp * (coeffs.y + 1.0)
    + value_end * lerp * coeffs.z
    + tangent_in_end * step_duration * lerp * coeffs.w
```

Comment (IQuick143) [reply crates/bevy_animation/src/gltf_curves.rs]: Oh right, I see.

Yeah this is most likely going to be a bunch faster. I did not realise you meant we can batch the coefficients like this (which should hopefully use SIMD under the hood too.)

File: crates/bevy_animation/src/animation_curves.rs
Comment (aecsocket) [inline crates/bevy_animation/src/animation_curves.rs:23]:
```suggestion
//! For instance, let's imagine that we want to use the `Vec3` output
```

File: crates/bevy_animation/src/animation_curves.rs
Comment (aecsocket) [inline crates/bevy_animation/src/animation_curves.rs:157]:
```suggestion
/// Here, the use of [`AnimatableKeyframeCurve`] creates a curve out of the given keyframe time-value
/// pairs, using the [`Animatable`] implementation of `f32` to interpolate between them. The
/// invocation of [`AnimatableCurve::from_curve`] with `FontSizeProperty` indicates that the `f32`
/// output from that curve is to be used to animate the font size of a `Text` component (as
/// configured above).
```

File: crates/bevy_math/src/curve/adaptors.rs
Comment (aecsocket) [inline crates/bevy_math/src/curve/adaptors.rs:459]: The name `ProductCurve` might be confused for the numerical product of the data from the two curves (at least when I read this type name initially, that's what I thought it was).

Since the fn is already `Curve::zip`, could this just be called `ZipCurve`?

File: crates/bevy_animation/src/gltf_curves.rs
Comment (aecsocket) [inline crates/bevy_animation/src/gltf_curves.rs:295]:
```suggestion
    #[error("number of values ({values_given}) is not divisible by {divisor}")]
```

File: crates/bevy_animation/src/gltf_curves.rs
Comment (aecsocket) [inline crates/bevy_animation/src/gltf_curves.rs:292]:
```suggestion
#[error("unable to construct a curve using this data")]
```

File: crates/bevy_animation/src/gltf_curves.rs
Comment (aecsocket) [inline crates/bevy_animation/src/gltf_curves.rs:308]:
```suggestion
    /// Create a new [`WideCubicKeyframeCurve`].
    ///
    /// An error will be returned if:
```
keeps the first line short and descriptive, so that when it appears as a listing in rustdoc, the "an error" part won't also be visible

File: examples/animation/animated_transform.rs
Comment (aecsocket) [inline examples/animation/animated_transform.rs:66]: nit: since this an example I'll be strict on following rust api guidelines, to encourage good practices
```suggestion
        .expect("failed to build translation curve"),
```

File: examples/animation/animated_ui.rs
Comment (aecsocket) [inline examples/animation/animated_ui.rs:90]:
```suggestion
            .expect("should be able to build translation curve because we pass in valid samples"),
```

File: examples/animation/animated_ui.rs
Comment (aecsocket) [inline examples/animation/animated_ui.rs:107]:
```suggestion
            .expect("should be able to build translation curve because we pass in valid samples"),
```

File: examples/animation/animated_transform.rs
Comment (aecsocket) [inline examples/animation/animated_transform.rs:66]: Under what condition could this curve fail to build? If we're passing it static data that we know is good, I feel like the `failed to build translation curve` message is almost misleading. As if this code even has a chance to fail.
```suggestion
        .expect("should be able to build translation curve because we pass in valid samples"),
```

File: examples/animation/animated_transform.rs
Comment (aecsocket) [inline examples/animation/animated_transform.rs:124]:
```suggestion
        .expect("should be able to build translation curve because we pass in valid samples"),
```

**Review by aecsocket — COMMENTED** [review]
LGTM, no problems with the logic that I can see, just some of the docs and text in general

File: crates/bevy_math/src/curve/adaptors.rs
Comment (mweatherley) [reply crates/bevy_math/src/curve/adaptors.rs]: Good idea. Done!

**Review by aecsocket — APPROVED** [review]

**Review by alice-i-cecile — APPROVED** [review]
Delightful. I really like the API, the general direction, and especially the detailed, conversational docs. Merging.

**Comment by alice-i-cecile** [comment]
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1708 if you'd like to help out.

**Pull request merged**
